### PR TITLE
Unrolling logical getindex and setindex for tuples

### PR DESCRIPTION
**Pull request by bramtayl**

### CODE REVIEW

File: base/tuple.jl [hunk 1]
```diff
@@ -31,6 +31,42 @@ function _setindex(v, i::Integer, first, tail...)
...
+getindex(into::Tuple{}, index::Tuple) = ()
+getindex(into::Tuple, index::Tuple) = begin
+    next = getindex(tail(into), tail(index))
+    ifelse(first(index), (first(into), next...), next)
+end
+
+setindex(old::Tuple{}, new, switch::Tuple{}) = ()
+setindex(old::Tuple{}, new, switch::Tuple) = ()
+setindex(old::Tuple, new, switch::Tuple{}) = old
+setindex(old::Tuple, new, switch::Tuple) =  begin
```
Comment (ararslan) [inline base/tuple.jl]: Please use long-form function definitions for functions that span multiple lines. It's also clearer to surround returned tuples in parentheses when one of the arguments has to be splatted.

In this case, the function doesn't actually need to span multiple lines, as the `first_switch` definition is unnecessary. In keeping with the Base style, this should be
```julia
setindex(old::Tuple, new, switch::Tuple) =
    (ifelse(first(switch), new, first(old)), setindex(tail(old), new, tail(switch))...)
```

File: test/tuple.jl [hunk 2]
```diff
@@ -116,6 +116,10 @@ end
     # @test_throws ArgumentError size((1,2), 2)
 end
 
+t1 = (1, "a", "b")
```
Comment (ararslan) [inline test/tuple.jl]: If these are only used in the test set you've added, they should be defined in that test set rather than at the top level.

**Review by ararslan — CHANGES_REQUESTED** [review]

File: base/tuple.jl [hunk 3]
```diff
@@ -31,6 +31,42 @@ function _setindex(v, i::Integer, first, tail...)
...
+    first_switch = first(switch)
+    ifelse(first_switch, new, first(old)),
+    setindex(
+        tail(old),
+        new,
+        tail(switch))...
+end
+
+setindex(old::Tuple{}, new::Tuple{}, switch::Tuple{}) = ()
```
Comment (JeffBezanson) [inline base/tuple.jl]: I don't think this behavior (doing something different when the value to assign is a `Tuple`) fits well with how tuples work --- a tuple is equally likely to contain any kind of value, including another tuple. The similar method for assigning arrays (`array1[inds] = array2`) I think is also a wart, but less harmful in practice due to how arrays tend to be used.

Comment (bramtayl) [reply base/tuple.jl]: Ok. Is there a better name to use for this behavior? Seems useful.

Comment (timholy) [reply base/tuple.jl]: Now that we have nice broadcasting of tuples, I think this is the same thing as `ifelse.(mask, input, default)`, which already works and generates efficient code:
```julia
julia> foo(mask, input, default) = ifelse.(mask, input, default)
foo (generic function with 1 method)

julia> @code_warntype foo((false, true), (1, 2), 0)
Variables:
  mask::Tuple{Bool,Bool}
  input::Tuple{Int64,Int64}
  default::Int64

Body:
  begin
      # meta: location broadcast.jl broadcast 434
      # meta: location broadcast.jl broadcast_c 335
      # meta: location broadcast.jl tuplebroadcast 337
      SSAValue(5) = input::Tuple{Int64,Int64}
      SSAValue(6) = default::Int64
      # meta: pop location
      # meta: pop location
      # meta: pop location
      return (Core.tuple)((Base.select_value)((Base.getfield)(mask::Tuple{Bool,Bool}, 1, true)::Bool, (Base.getfield)(SSAValue(5), 1, true)::Int64, SSAValue(6))::Int64, (Base.select_value)((Base.getfield)(mask::Tuple{Bool,Bool}, 2, true)::Bool, (Base.getfield)(SSAValue(5), 2, true)::Int64, SSAValue(6))::Int64)::Tuple{Int64,Int64}
  end::Tuple{Int64,Int64}
```
I don't doubt that there are many places in Base that could be cleaned up by using this.

Comment (bramtayl) [reply base/tuple.jl]: Oh yeah, that is much nicer for methods of setindex where the replacement value is a scalar. Still need something for slotting in a tuple of replacement values?

**Pull request closed without merging**